<!DOCTYPE html>

<html lang="en" data-content_root="../">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Taxonomy databases &#8212; ETE Toolkit 4.0.0-beta2 documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=4f649999" />
    <link rel="stylesheet" type="text/css" href="../_static/alabaster.css?v=039e1c02" />
    <script src="../_static/documentation_options.js?v=382e6e0c"></script>
    <script src="../_static/doctools.js?v=888ff710"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <link rel="icon" href="../_static/favicon.ico"/>
    <link rel="author" title="About these documents" href="../about.html" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Reference Guide" href="../reference/index.html" />
    <link rel="prev" title="Programmable tree drawing" href="tutorial_drawing.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="taxonomy-databases">
<h1><a class="toc-backref" href="#id1" role="doc-backlink">Taxonomy databases</a><a class="headerlink" href="#taxonomy-databases" title="Link to this heading">¶</a></h1>
<nav class="contents" id="contents">
<p class="topic-title">Contents</p>
<ul class="simple">
<li><p><a class="reference internal" href="#taxonomy-databases" id="id1">Taxonomy databases</a></p>
<ul>
<li><p><a class="reference internal" href="#overview" id="id2">Overview</a></p></li>
<li><p><a class="reference internal" href="#differences-between-ncbi-and-gtdb-taxonomies-in-ete4" id="id3">Differences between NCBI and GTDB taxonomies in ETE4</a></p></li>
<li><p><a class="reference internal" href="#setting-up-local-copies-of-the-ncbi-and-gtdb-taxonomy-databases" id="id4">Setting up local copies of the NCBI and GTDB taxonomy databases</a></p></li>
<li><p><a class="reference internal" href="#getting-taxid-information" id="id5">Getting taxid information</a></p>
<ul>
<li><p><a class="reference internal" href="#ncbi-taxonomy" id="id6">NCBI taxonomy</a></p></li>
<li><p><a class="reference internal" href="#gtdb-taxonomy" id="id7">GTDB taxonomy</a></p></li>
</ul>
</li>
<li><p><a class="reference internal" href="#getting-descendant-taxa" id="id8">Getting descendant taxa</a></p></li>
<li><p><a class="reference internal" href="#getting-species-tree-topology" id="id9">Getting species tree topology</a></p></li>
<li><p><a class="reference internal" href="#automatic-tree-annotation-using-ncbi-gtdb-taxonomy" id="id10">Automatic tree annotation using NCBI/GTDB taxonomy</a></p></li>
</ul>
</li>
</ul>
</nav>
<section id="overview">
<h2><a class="toc-backref" href="#id2" role="doc-backlink">Overview</a><a class="headerlink" href="#overview" title="Link to this heading">¶</a></h2>
<p>ETE4 contains the <em>ncbi_taxonomy</em> and <em>gtdb_taxonomy</em> modules which
provide utilities to efficiently query a local copy of the NCBI or
GTDB taxonomy databases. The classes <a class="reference internal" href="../reference/reference_taxonomy.html#ete4.NCBITaxa" title="ete4.NCBITaxa"><code class="xref py py-class docutils literal notranslate"><span class="pre">NCBITaxa</span></code></a> and
<a class="reference internal" href="../reference/reference_taxonomy.html#ete4.GTDBTaxa" title="ete4.GTDBTaxa"><code class="xref py py-class docutils literal notranslate"><span class="pre">GTDBTaxa</span></code></a> offer methods to convert from taxid to names (and
vice versa), to fetch pruned topologies connecting a given set of
species, or to download rank, names and lineage track information.</p>
<p>It is also fully integrated with <a class="reference internal" href="../reference/reference_phylo.html#ete4.PhyloTree" title="ete4.PhyloTree"><code class="xref py py-class docutils literal notranslate"><span class="pre">PhyloTree</span></code></a> instances through
the <a class="reference internal" href="../reference/reference_phylo.html#ete4.PhyloTree.annotate_ncbi_taxa" title="ete4.PhyloTree.annotate_ncbi_taxa"><code class="xref py py-func docutils literal notranslate"><span class="pre">annotate_ncbi_taxa()</span></code></a> and
<a class="reference internal" href="../reference/reference_phylo.html#ete4.PhyloTree.annotate_gtdb_taxa" title="ete4.PhyloTree.annotate_gtdb_taxa"><code class="xref py py-func docutils literal notranslate"><span class="pre">annotate_gtdb_taxa()</span></code></a> methods.</p>
</section>
<section id="differences-between-ncbi-and-gtdb-taxonomies-in-ete4">
<h2><a class="toc-backref" href="#id3" role="doc-backlink">Differences between NCBI and GTDB taxonomies in ETE4</a><a class="headerlink" href="#differences-between-ncbi-and-gtdb-taxonomies-in-ete4" title="Link to this heading">¶</a></h2>
<p>The NCBI taxonomy database is a comprehensive resource for organism
names and classifications.It is updated daily and offers multiple
access points including a web portal, an FTP server. The database
releases its data in a package called “taxdump.tar.gz” which contains
several .dmp files.</p>
<p>Taxon in NCBI taxonomyis usually a numeric identifier, commonly
representing taxa (“TaxID”), but it can also signify other entities
like genetic codes or citations, such as 9606 represents Homo Sapiens.</p>
<p>On the other hand, GTDB taxonomy is distributed as simple text files,
uses a genome-based approach for classification, and the identifiers
are usually specific to genomes rather than taxa.</p>
<p>Since ETE Toolkit version 3, ETE parses taxdump file and stores it in
a local sqlite database to fullfill the methods in ncbi_taxonomy
module. We applied the same strategy to GTDBTaxa. While the original
GTDB taxonomy data differs from NCBI taxonomy files, a conversion step
is essential for integration.</p>
<p>To integrate GTDB into the ETE Toolkit v4, a conversion process is
necessary. A third-party script
(<a class="reference external" href="https://github.com/nick-youngblut/gtdb_to_taxdump">https://github.com/nick-youngblut/gtdb_to_taxdump</a>) is employed to
convert the GTDB taxonomy to the NCBI-like taxdump format. We already
preprared GTDB taxonomy dump file from different releases version and
store in
<a class="reference external" href="https://github.com/etetoolkit/ete-data/tree/main/gtdb_taxonomy">https://github.com/etetoolkit/ete-data/tree/main/gtdb_taxonomy</a>.</p>
</section>
<section id="setting-up-local-copies-of-the-ncbi-and-gtdb-taxonomy-databases">
<h2><a class="toc-backref" href="#id4" role="doc-backlink">Setting up local copies of the NCBI and GTDB taxonomy databases</a><a class="headerlink" href="#setting-up-local-copies-of-the-ncbi-and-gtdb-taxonomy-databases" title="Link to this heading">¶</a></h2>
<p>The first time you attempt to use NCBITaxa or GTDBTaxa, ETE will
detect that your local database is empty and will attempt to download
the latest taxonomy database (NCBI ~600MB, GTDB ~72MB) and will store
a parsed version of it in <cite>~/.local/share/ete/</cite> by default. All future
imports of NCBITaxa or GTDBTaxa will detect the local database and
will skip this step.</p>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Load NCBI module</span>
<span class="kn">from</span> <span class="nn">ete4</span> <span class="kn">import</span> <span class="n">NCBITaxa</span>
<span class="n">ncbi</span> <span class="o">=</span> <span class="n">NCBITaxa</span><span class="p">()</span>
<span class="n">ncbi</span><span class="o">.</span><span class="n">update_taxonomy_database</span><span class="p">()</span>

<span class="c1"># Load GTDB module</span>
<span class="kn">from</span> <span class="nn">ete4</span> <span class="kn">import</span> <span class="n">GTDBTaxa</span>
<span class="n">gtdb</span> <span class="o">=</span> <span class="n">GTDBTaxa</span><span class="p">()</span>
<span class="n">gtdb</span><span class="o">.</span><span class="n">update_taxonomy_database</span><span class="p">()</span>

<span class="c1"># Load GTDB module with specific release version</span>
<span class="kn">from</span> <span class="nn">ete4</span> <span class="kn">import</span> <span class="n">GTDBTaxa</span>
<span class="n">gtdb</span> <span class="o">=</span> <span class="n">GTDBTaxa</span><span class="p">()</span>

<span class="c1"># latest release updated in https://github.com/dengzq1234/ete-data/tree/main/gtdb_taxonomy</span>
<span class="n">gtdb</span><span class="o">.</span><span class="n">update_taxonomy_database</span><span class="p">()</span>
<span class="c1"># or</span>
<span class="n">gtdb</span><span class="o">.</span><span class="n">update_taxonomy_database</span><span class="p">(</span><span class="s2">&quot;gtdbdump.tar.gz&quot;</span><span class="p">)</span>

<span class="c1"># update with custom release 202</span>
<span class="n">gtdb</span><span class="o">.</span><span class="n">update_taxonomy_database</span><span class="p">(</span><span class="s1">&#39;gtdb202dump.tar.gz&#39;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="getting-taxid-information">
<h2><a class="toc-backref" href="#id5" role="doc-backlink">Getting taxid information</a><a class="headerlink" href="#getting-taxid-information" title="Link to this heading">¶</a></h2>
<section id="ncbi-taxonomy">
<h3><a class="toc-backref" href="#id6" role="doc-backlink">NCBI taxonomy</a><a class="headerlink" href="#ncbi-taxonomy" title="Link to this heading">¶</a></h3>
<p>You can fetch species names, ranks and linage track information for
your taxids using the following methods:</p>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="../reference/reference_taxonomy.html#ete4.NCBITaxa.get_rank" title="ete4.NCBITaxa.get_rank"><code class="xref py py-obj docutils literal notranslate"><span class="pre">NCBITaxa.get_rank</span></code></a>(taxids)</p></td>
<td><p>Return dict with NCBI taxonomy ranks for each list of taxids.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="../reference/reference_taxonomy.html#ete4.NCBITaxa.get_lineage" title="ete4.NCBITaxa.get_lineage"><code class="xref py py-obj docutils literal notranslate"><span class="pre">NCBITaxa.get_lineage</span></code></a>(taxid)</p></td>
<td><p>Return lineage track corresponding to the given taxid.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="../reference/reference_taxonomy.html#ete4.NCBITaxa.get_taxid_translator" title="ete4.NCBITaxa.get_taxid_translator"><code class="xref py py-obj docutils literal notranslate"><span class="pre">NCBITaxa.get_taxid_translator</span></code></a>(taxids[, ...])</p></td>
<td><p>Return dict with the scientific names corresponding to the taxids.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="../reference/reference_taxonomy.html#ete4.NCBITaxa.get_name_translator" title="ete4.NCBITaxa.get_name_translator"><code class="xref py py-obj docutils literal notranslate"><span class="pre">NCBITaxa.get_name_translator</span></code></a>(names)</p></td>
<td><p>Return dict with taxids corresponding to the given scientific names.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="../reference/reference_taxonomy.html#ete4.NCBITaxa.translate_to_names" title="ete4.NCBITaxa.translate_to_names"><code class="xref py py-obj docutils literal notranslate"><span class="pre">NCBITaxa.translate_to_names</span></code></a>(taxids)</p></td>
<td><p>Return list of scientific names corresponding to taxids.</p></td>
</tr>
</tbody>
</table>
<p>The so called get-translator functions will return a dictionary
converting between taxids and species names. Either species or linage
names/taxids are accepted as input.</p>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">ete4</span> <span class="kn">import</span> <span class="n">NCBITaxa</span>
<span class="n">ncbi</span> <span class="o">=</span> <span class="n">NCBITaxa</span><span class="p">()</span>
<span class="n">taxid2name</span> <span class="o">=</span> <span class="n">ncbi</span><span class="o">.</span><span class="n">get_taxid_translator</span><span class="p">([</span><span class="mi">9606</span><span class="p">,</span> <span class="mi">9443</span><span class="p">])</span>
<span class="nb">print</span><span class="p">(</span><span class="n">taxid2name</span><span class="p">)</span>
<span class="c1"># {9443: &#39;Primates&#39;, 9606: &#39;Homo sapiens&#39;}</span>

<span class="n">name2taxid</span> <span class="o">=</span> <span class="n">ncbi</span><span class="o">.</span><span class="n">get_name_translator</span><span class="p">([</span><span class="s1">&#39;Homo sapiens&#39;</span><span class="p">,</span> <span class="s1">&#39;primates&#39;</span><span class="p">])</span>
<span class="nb">print</span><span class="p">(</span><span class="n">name2taxid</span><span class="p">)</span>
<span class="c1"># {&#39;Homo sapiens&#39;: [9606], &#39;primates&#39;: [9443]}</span>

<span class="c1"># when the same name points to several taxa, all taxids are returned</span>
<span class="n">name2taxid</span> <span class="o">=</span> <span class="n">ncbi</span><span class="o">.</span><span class="n">get_name_translator</span><span class="p">([</span><span class="s1">&#39;Bacteria&#39;</span><span class="p">])</span>
<span class="nb">print</span><span class="p">(</span><span class="n">name2taxid</span><span class="p">)</span>
<span class="c1"># {&#39;Bacteria&#39;: [2, 629395]}</span>
</pre></div>
</div>
<p>Other functions allow to extract further information using taxid
numbers as a query.</p>
<p>Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">ete4</span> <span class="kn">import</span> <span class="n">NCBITaxa</span>
<span class="n">ncbi</span> <span class="o">=</span> <span class="n">NCBITaxa</span><span class="p">()</span>

<span class="nb">print</span><span class="p">(</span><span class="n">ncbi</span><span class="o">.</span><span class="n">get_rank</span><span class="p">([</span><span class="mi">9606</span><span class="p">,</span> <span class="mi">9443</span><span class="p">]))</span>
<span class="c1"># {9443: &#39;order&#39;, 9606: &#39;species&#39;}</span>

<span class="nb">print</span><span class="p">(</span><span class="n">ncbi</span><span class="o">.</span><span class="n">get_lineage</span><span class="p">(</span><span class="mi">9606</span><span class="p">))</span>
<span class="c1"># [1, 131567, 2759, 33154, 33208, 6072, 33213, 33511, 7711, 89593, 7742,</span>
<span class="c1"># 7776, 117570, 117571, 8287, 1338369, 32523, 32524, 40674, 32525, 9347,</span>
<span class="c1"># 1437010, 314146, 9443, 376913, 314293, 9526, 314295, 9604, 207598, 9605,</span>
<span class="c1"># 9606]</span>
</pre></div>
</div>
<p>Example combining all at once:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">ete4</span> <span class="kn">import</span> <span class="n">NCBITaxa</span>
<span class="n">ncbi</span> <span class="o">=</span> <span class="n">NCBITaxa</span><span class="p">()</span>

<span class="n">lineage</span> <span class="o">=</span> <span class="n">ncbi</span><span class="o">.</span><span class="n">get_lineage</span><span class="p">(</span><span class="mi">9606</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">lineage</span><span class="p">)</span>

<span class="c1"># [1, 131567, 2759, 33154, 33208, 6072, 33213, 33511, 7711, 89593, 7742,</span>
<span class="c1"># 7776, 117570, 117571, 8287, 1338369, 32523, 32524, 40674, 32525, 9347,</span>
<span class="c1"># 1437010, 314146, 9443, 376913, 314293, 9526, 314295, 9604, 207598, 9605,</span>
<span class="c1"># 9606]</span>

<span class="n">names</span> <span class="o">=</span> <span class="n">ncbi</span><span class="o">.</span><span class="n">get_taxid_translator</span><span class="p">(</span><span class="n">lineage</span><span class="p">)</span>
<span class="nb">print</span><span class="p">([</span><span class="n">names</span><span class="p">[</span><span class="n">taxid</span><span class="p">]</span> <span class="k">for</span> <span class="n">taxid</span> <span class="ow">in</span> <span class="n">lineage</span><span class="p">])</span>

<span class="c1"># [&#39;root&#39;, &#39;cellular organisms&#39;, &#39;Eukaryota&#39;, &#39;Opisthokonta&#39;, &#39;Metazoa&#39;,</span>
<span class="c1"># &#39;Eumetazoa&#39;, &#39;Bilateria&#39;, &#39;Deuterostomia&#39;, &#39;Chordata&#39;, &#39;Craniata&#39;,</span>
<span class="c1"># &#39;Vertebrata&#39;, &#39;Gnathostomata&#39;, &#39;Teleostomi&#39;, &#39;Euteleostomi&#39;,</span>
<span class="c1"># &#39;Sarcopterygii&#39;, &#39;Dipnotetrapodomorpha&#39;, &#39;Tetrapoda&#39;, &#39;Amniota&#39;,</span>
<span class="c1"># &#39;Mammalia&#39;, &#39;Theria&#39;, &#39;Eutheria&#39;, &#39;Boreoeutheria&#39;, &#39;Euarchontoglires&#39;,</span>
<span class="c1"># &#39;Primates&#39;, &#39;Haplorrhini&#39;, &#39;Simiiformes&#39;, &#39;Catarrhini&#39;, &#39;Hominoidea&#39;,</span>
<span class="c1"># &#39;Hominidae&#39;, &#39;Homininae&#39;, &#39;Homo&#39;, &#39;Homo sapiens&#39;]</span>
</pre></div>
</div>
</section>
<section id="gtdb-taxonomy">
<h3><a class="toc-backref" href="#id7" role="doc-backlink">GTDB taxonomy</a><a class="headerlink" href="#gtdb-taxonomy" title="Link to this heading">¶</a></h3>
<p>In the NCBI taxonomy database, each species is assigned a unique numeric taxid.
For example, the taxid 9606 refers to Homo sapiens. These taxids serve as
essential keys for tracking lineages within the database.</p>
<p>However, the GTDB database doesn’t originally offer numeric taxids like NCBI
does. In the GTDBTaxa module, we’ve introduced taxids for each species to
facilitate lineage tracking. These taxids, while not officially recognized in
the GTDB database, serve as convenient keys. They help in connecting the lineage
and taxonomic ranks within the local database, making it easier for users to
fetch and relate taxonomic information.</p>
<p>Like NCBITaxa, GTDBTaxa contains similar methods:</p>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="../reference/reference_taxonomy.html#ete4.GTDBTaxa.get_rank" title="ete4.GTDBTaxa.get_rank"><code class="xref py py-obj docutils literal notranslate"><span class="pre">GTDBTaxa.get_rank</span></code></a>(taxids)</p></td>
<td><p>Return dictionary converting taxids to their GTDB taxonomy rank.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="../reference/reference_taxonomy.html#ete4.GTDBTaxa.get_lineage" title="ete4.GTDBTaxa.get_lineage"><code class="xref py py-obj docutils literal notranslate"><span class="pre">GTDBTaxa.get_lineage</span></code></a>(taxid)</p></td>
<td><p>Given a valid taxid number, return its corresponding lineage track as a hierarchically sorted list of parent taxids.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="../reference/reference_taxonomy.html#ete4.GTDBTaxa.get_taxid_translator" title="ete4.GTDBTaxa.get_taxid_translator"><code class="xref py py-obj docutils literal notranslate"><span class="pre">GTDBTaxa.get_taxid_translator</span></code></a>(taxids[, ...])</p></td>
<td><p>Given a list of taxids, returns a dictionary with their corresponding scientific names.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="../reference/reference_taxonomy.html#ete4.GTDBTaxa.get_name_translator" title="ete4.GTDBTaxa.get_name_translator"><code class="xref py py-obj docutils literal notranslate"><span class="pre">GTDBTaxa.get_name_translator</span></code></a>(names)</p></td>
<td><p>Given a list of taxid scientific names, returns a dictionary translating them into their corresponding taxids.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="../reference/reference_taxonomy.html#ete4.GTDBTaxa.translate_to_names" title="ete4.GTDBTaxa.translate_to_names"><code class="xref py py-obj docutils literal notranslate"><span class="pre">GTDBTaxa.translate_to_names</span></code></a>(taxids)</p></td>
<td><p>Given a list of taxid numbers, returns another list with their corresponding scientific names.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="../reference/reference_taxonomy.html#ete4.GTDBTaxa.get_name_lineage" title="ete4.GTDBTaxa.get_name_lineage"><code class="xref py py-obj docutils literal notranslate"><span class="pre">GTDBTaxa.get_name_lineage</span></code></a>(taxnames)</p></td>
<td><p>Given a valid taxname, return its corresponding lineage track as a hierarchically sorted list of parent taxnames.</p></td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="getting-descendant-taxa">
<h2><a class="toc-backref" href="#id8" role="doc-backlink">Getting descendant taxa</a><a class="headerlink" href="#getting-descendant-taxa" title="Link to this heading">¶</a></h2>
<p>Given a taxid or a taxa name from an internal node in the NCBI/GTDB taxonomy tree,
their descendants can be retrieved as follows:</p>
<p>NCBI taxonomy example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">ete4</span> <span class="kn">import</span> <span class="n">NCBITaxa</span>
<span class="n">ncbi</span> <span class="o">=</span> <span class="n">NCBITaxa</span><span class="p">()</span>

<span class="n">descendants</span> <span class="o">=</span> <span class="n">ncbi</span><span class="o">.</span><span class="n">get_descendant_taxa</span><span class="p">(</span><span class="s1">&#39;Homo&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">ncbi</span><span class="o">.</span><span class="n">translate_to_names</span><span class="p">(</span><span class="n">descendants</span><span class="p">))</span>

<span class="c1"># [&#39;Homo heidelbergensis&#39;, &#39;Homo sapiens ssp. Denisova&#39;,</span>
<span class="c1"># &#39;Homo sapiens neanderthalensis&#39;]</span>

<span class="c1"># You can easily ignore subspecies, so only taxa labeled as &quot;species&quot; will be reported:</span>
<span class="n">descendants</span> <span class="o">=</span> <span class="n">ncbi</span><span class="o">.</span><span class="n">get_descendant_taxa</span><span class="p">(</span><span class="s1">&#39;Homo&#39;</span><span class="p">,</span> <span class="n">collapse_subspecies</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">ncbi</span><span class="o">.</span><span class="n">translate_to_names</span><span class="p">(</span><span class="n">descendants</span><span class="p">))</span>

<span class="c1"># [&#39;Homo sapiens&#39;, &#39;Homo heidelbergensis&#39;]</span>

<span class="c1"># or even returned as an annotated tree</span>
<span class="n">tree</span> <span class="o">=</span> <span class="n">ncbi</span><span class="o">.</span><span class="n">get_descendant_taxa</span><span class="p">(</span><span class="s1">&#39;Homo&#39;</span><span class="p">,</span> <span class="n">collapse_subspecies</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">return_tree</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">to_str</span><span class="p">(</span><span class="n">props</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;sci_name&#39;</span><span class="p">,</span><span class="s1">&#39;taxid&#39;</span><span class="p">]))</span>
<span class="c1">#            ╭╴environmental samples,2665952╶╌╴Homo sapiens environmental sample,2665953</span>
<span class="c1">#            │</span>
<span class="c1">#            ├╴Homo sapiens,9606</span>
<span class="c1"># ╴Homo,9605╶┤</span>
<span class="c1">#            ├╴Homo heidelbergensis,1425170</span>
<span class="c1">#            │</span>
<span class="c1">#            ╰╴unclassified Homo,2813598╶╌╴Homo sp.,2813599</span>
</pre></div>
</div>
<p>GTDB taxonomy example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">ete4</span> <span class="kn">import</span> <span class="n">GTDBTaxa</span>
<span class="n">gtdb</span> <span class="o">=</span> <span class="n">GTDBTaxa</span><span class="p">()</span>
<span class="n">descendants</span> <span class="o">=</span> <span class="n">gtdb</span><span class="o">.</span><span class="n">get_descendant_taxa</span><span class="p">(</span><span class="s1">&#39;f__Thorarchaeaceae&#39;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">descendants</span><span class="p">)</span>
<span class="c1"># [&#39;GB_GCA_003662765.1&#39;, &#39;GB_GCA_003662805.1&#39;, ..., &#39;GB_GCA_013138615.1&#39;]</span>

<span class="c1"># Ignore subspecies, so only taxa labeled as &quot;species&quot; will be reported.</span>
<span class="n">descendants</span> <span class="o">=</span> <span class="n">gtdb</span><span class="o">.</span><span class="n">get_descendant_taxa</span><span class="p">(</span><span class="s1">&#39;f__Thorarchaeaceae&#39;</span><span class="p">,</span> <span class="n">collapse_subspecies</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">descendants</span><span class="p">)</span>
<span class="c1"># [&#39;s__MP8T-1 sp002825535&#39;, &#39;s__MP8T-1 sp003345545&#39;, ..., &#39;s__TEKIR-12S sp004524435&#39;]</span>

<span class="c1"># Returned as an annotated tree.</span>
<span class="n">descendants</span> <span class="o">=</span> <span class="n">gtdb</span><span class="o">.</span><span class="n">get_descendant_taxa</span><span class="p">(</span><span class="s1">&#39;f__Thorarchaeaceae&#39;</span><span class="p">,</span> <span class="n">collapse_subspecies</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">return_tree</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">descendants</span><span class="o">.</span><span class="n">to_str</span><span class="p">(</span><span class="n">props</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;sci_name&#39;</span><span class="p">,</span><span class="s1">&#39;rank&#39;</span><span class="p">]))</span>
<span class="c1">#                                ╭╴s__MP8T-1 sp002825535,species</span>
<span class="c1">#                                │</span>
<span class="c1">#                                ├╴s__MP8T-1 sp003345545,species</span>
<span class="c1">#                                │</span>
<span class="c1">#              ╭╴g__MP8T-1,genus╶┼╴s__MP8T-1 sp002825465,species</span>
<span class="c1">#              │                 │</span>
<span class="c1">#              │                 ├╴s__MP8T-1 sp004524565,species</span>
<span class="c1">#              │                 │</span>
<span class="c1">#              │                 ╰╴s__MP8T-1 sp004524595,species</span>
<span class="c1">#              │</span>
<span class="c1">#              │                   ╭╴s__SMTZ1-83 sp011364985,species</span>
<span class="c1">#              │                   │</span>
<span class="c1">#              ├╴g__SMTZ1-83,genus╶┼╴s__SMTZ1-83 sp011365025,species</span>
<span class="c1">#              │                   │</span>
<span class="c1">#              │                   ╰╴s__SMTZ1-83 sp001563325,species</span>
<span class="c1">#              │</span>
<span class="c1">#              ├╴g__TEKIR-14,genus╶╌╴s__TEKIR-14 sp004524445,species</span>
<span class="c1">#              │</span>
<span class="c1">#              ├╴g__SHMX01,genus╶╌╴s__SHMX01 sp008080745,species</span>
<span class="c1">#              │</span>
<span class="c1">#              │               ╭╴s__OWC5 sp003345595,species</span>
<span class="c1">#              ├╴g__OWC5,genus╶┤</span>
<span class="c1"># ╴f__Tho[...]╶┤               ╰╴s__OWC5 sp003345555,species</span>
<span class="c1">#              │</span>
<span class="c1">#              ├╴g__JACAEL01,genus╶╌╴s__JACAEL01 sp013388835,species</span>
<span class="c1">#              │</span>
<span class="c1">#              ├╴g__B65-G9,genus╶╌╴s__B65-G9 sp003662765,species</span>
<span class="c1">#              │</span>
<span class="c1">#              │                   ╭╴s__SMTZ1-45 sp001563335,species</span>
<span class="c1">#              │                   │</span>
<span class="c1">#              │                   ├╴s__SMTZ1-45 sp011364905,species</span>
<span class="c1">#              │                   │</span>
<span class="c1">#              ├╴g__SMTZ1-45,genus╶┼╴s__SMTZ1-45 sp001940705,species</span>
<span class="c1">#              │                   │</span>
<span class="c1">#              │                   ├╴s__SMTZ1-45 sp004376265,species</span>
<span class="c1">#              │                   │</span>
<span class="c1">#              │                   ╰╴s__SMTZ1-45 sp002825515,species</span>
<span class="c1">#              │</span>
<span class="c1">#              ├╴g__WTCK01,genus╶╌╴s__WTCK01 sp013138615,species</span>
<span class="c1">#              │</span>
<span class="c1">#              ╰╴g__TEKIR-12S,genus╶╌╴s__TEKIR-12S sp004524435,species</span>
</pre></div>
</div>
</section>
<section id="getting-species-tree-topology">
<h2><a class="toc-backref" href="#id9" role="doc-backlink">Getting species tree topology</a><a class="headerlink" href="#getting-species-tree-topology" title="Link to this heading">¶</a></h2>
<p>Getting the taxonomy tree for a given set of species is one of the
most useful ways to get all information at once. The methods
<a class="reference internal" href="../reference/reference_taxonomy.html#ete4.NCBITaxa.get_topology" title="ete4.NCBITaxa.get_topology"><code class="xref py py-func docutils literal notranslate"><span class="pre">NCBITaxa.get_topology()</span></code></a> or <a class="reference internal" href="../reference/reference_taxonomy.html#ete4.GTDBTaxa.get_topology" title="ete4.GTDBTaxa.get_topology"><code class="xref py py-func docutils literal notranslate"><span class="pre">GTDBTaxa.get_topology()</span></code></a> allow
to query your local NCBI/GTDB database and extract the smallest tree
that connects all your query taxids. It returns a normal ETE tree in
which all nodes, internal or leaves, are annotated for lineage,
scientific names, ranks, and so on.</p>
<p>NCBI taxonomy example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">ete4</span> <span class="kn">import</span> <span class="n">NCBITaxa</span>
<span class="n">ncbi</span> <span class="o">=</span> <span class="n">NCBITaxa</span><span class="p">()</span>

<span class="n">tree</span> <span class="o">=</span> <span class="n">ncbi</span><span class="o">.</span><span class="n">get_topology</span><span class="p">([</span><span class="mi">9606</span><span class="p">,</span> <span class="mi">9598</span><span class="p">,</span> <span class="mi">10090</span><span class="p">,</span> <span class="mi">7707</span><span class="p">,</span> <span class="mi">8782</span><span class="p">])</span>

<span class="nb">print</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">to_str</span><span class="p">(</span><span class="n">props</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;sci_name&quot;</span><span class="p">,</span> <span class="s2">&quot;rank&quot;</span><span class="p">]))</span>
<span class="c1">#                      ╭╴Dendrochirotida,order</span>
<span class="c1">#                      │</span>
<span class="c1">#                      │                                                                   ╭╴Homo sapiens,species</span>
<span class="c1"># ╴Deuterostomia,clade╶┤                                             ╭╴Homininae,subfamily╶┤</span>
<span class="c1">#                      │               ╭╴Euarchontoglires,superorder╶┤                     ╰╴Pan troglodytes,species</span>
<span class="c1">#                      │               │                             │</span>
<span class="c1">#                      ╰╴Amniota,clade╶┤                             ╰╴Mus musculus,species</span>
<span class="c1">#                                      │</span>
<span class="c1">#                                      ╰╴Aves,class</span>

<span class="c1"># All intermediate nodes connecting the species can also be kept in the tree.</span>
<span class="n">tree</span> <span class="o">=</span> <span class="n">ncbi</span><span class="o">.</span><span class="n">get_topology</span><span class="p">([</span><span class="mi">2</span><span class="p">,</span> <span class="mi">33208</span><span class="p">],</span> <span class="n">intermediate_nodes</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">to_str</span><span class="p">(</span><span class="n">props</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;sci_name&quot;</span><span class="p">]))</span>
<span class="c1">#                     ╭╴Eukaryota╶╌╴Opisthokonta╶╌╴Metazoa</span>
<span class="c1"># ╴cellular organisms╶┤</span>
<span class="c1">#                     ╰╴Bacteria</span>
</pre></div>
</div>
<p>GTDB taxonomy example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">ete4</span> <span class="kn">import</span> <span class="n">GTDBTaxa</span>
<span class="n">gtdb</span> <span class="o">=</span> <span class="n">GTDBTaxa</span><span class="p">()</span>

<span class="n">tree</span> <span class="o">=</span> <span class="n">gtdb</span><span class="o">.</span><span class="n">get_topology</span><span class="p">([</span><span class="s2">&quot;p__Huberarchaeota&quot;</span><span class="p">,</span> <span class="s2">&quot;o__Peptococcales&quot;</span><span class="p">,</span> <span class="s2">&quot;f__Korarchaeaceae&quot;</span><span class="p">])</span>
<span class="nb">print</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">to_str</span><span class="p">(</span><span class="n">props</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;sci_name&#39;</span><span class="p">,</span> <span class="s1">&#39;rank&#39;</span><span class="p">]))</span>
<span class="c1">#                                         ╭╴p__Huberarchaeota,phylum</span>
<span class="c1">#               ╭╴d__Archaea,superkingdom╶┤</span>
<span class="c1"># ╴root,no rank╶┤                         ╰╴f__Korarchaeaceae,family</span>
<span class="c1">#               │</span>
<span class="c1">#               ╰╴o__Peptococcales,order</span>

<span class="c1"># All intermediate nodes connecting the species can also be kept in the tree.</span>
<span class="n">tree</span> <span class="o">=</span> <span class="n">gtdb</span><span class="o">.</span><span class="n">get_topology</span><span class="p">([</span><span class="s2">&quot;p__Huberarchaeota&quot;</span><span class="p">,</span> <span class="s2">&quot;o__Peptococcales&quot;</span><span class="p">,</span> <span class="s2">&quot;f__Korarchaeaceae&quot;</span><span class="p">],</span> <span class="n">intermediate_nodes</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">collapse_subspecies</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">annotate</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">to_str</span><span class="p">(</span><span class="n">props</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;sci_name&#39;</span><span class="p">,</span> <span class="s1">&#39;rank&#39;</span><span class="p">]))</span>
<span class="c1">#                                         ╭╴p__Huberarchaeota,phylum</span>
<span class="c1">#               ╭╴d__Archaea,superkingdom╶┤</span>
<span class="c1"># ╴root,no rank╶┤                         ╰╴p__Thermoproteota,phylum╶╌╴c__Korarchaeia,class╶╌╴o__Korarchaeales,order╶╌╴f__Korarchaeaceae,family</span>
<span class="c1">#               │</span>
<span class="c1">#               ╰╴d__Bacteria,superkingdom╶╌╴p__Firmicutes_B,phylum╶╌╴c__Peptococcia,class╶╌╴o__Peptococcales,order</span>
</pre></div>
</div>
</section>
<section id="automatic-tree-annotation-using-ncbi-gtdb-taxonomy">
<h2><a class="toc-backref" href="#id10" role="doc-backlink">Automatic tree annotation using NCBI/GTDB taxonomy</a><a class="headerlink" href="#automatic-tree-annotation-using-ncbi-gtdb-taxonomy" title="Link to this heading">¶</a></h2>
<p>NCBI/GTDB taxonomy annotation consists of adding additional
information to any internal or leaf node in a tree. Only a property
containing the taxid associated to each node is required for the nodes
in the query tree. The annotation process will add the following
features to the nodes:</p>
<ul class="simple">
<li><p>sci_name</p></li>
<li><p>taxid</p></li>
<li><p>named_lineage</p></li>
<li><p>lineage</p></li>
<li><p>rank</p></li>
</ul>
<p>Note that, for internal nodes, taxid can be automatically inferred
based on their sibling nodes. The easiest way to annotate a tree is to
use a PhyloTree instance where the species name attribute is
transparently used as the taxid attribute. Note that the
<a class="reference internal" href="../reference/reference_phylo.html#ete4.PhyloTree.annotate_ncbi_taxa" title="ete4.PhyloTree.annotate_ncbi_taxa"><code class="xref py py-func docutils literal notranslate"><span class="pre">annotate_ncbi_taxa()</span></code></a> or
<a class="reference internal" href="../reference/reference_phylo.html#ete4.PhyloTree.annotate_gtdb_taxa" title="ete4.PhyloTree.annotate_gtdb_taxa"><code class="xref py py-func docutils literal notranslate"><span class="pre">annotate_gtdb_taxa()</span></code></a> function will also return the
used name, lineage and rank translators.</p>
<p>Remember that species names in PhyloTree instances are automatically
extracted from leaf names. The parsing method can be easily adapted to
any formatting:</p>
<p>NCBI taxonomy example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">ete4</span> <span class="kn">import</span> <span class="n">PhyloTree</span>

<span class="c1"># Load the whole leaf name as species taxid.</span>
<span class="n">tree</span> <span class="o">=</span> <span class="n">PhyloTree</span><span class="p">(</span><span class="s1">&#39;((9606, 9598), 10090);&#39;</span><span class="p">,</span> <span class="n">sp_naming_function</span><span class="o">=</span><span class="k">lambda</span> <span class="n">name</span><span class="p">:</span> <span class="n">name</span><span class="p">)</span>
<span class="n">tax2names</span><span class="p">,</span> <span class="n">tax2lineages</span><span class="p">,</span> <span class="n">tax2rank</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">annotate_ncbi_taxa</span><span class="p">()</span>

<span class="c1"># Split names by &#39;|&#39; and return the first part as the species taxid.</span>
<span class="n">tree</span> <span class="o">=</span> <span class="n">PhyloTree</span><span class="p">(</span><span class="s1">&#39;((9606|protA, 9598|protA), 10090|protB);&#39;</span><span class="p">,</span> <span class="n">sp_naming_function</span><span class="o">=</span><span class="k">lambda</span> <span class="n">name</span><span class="p">:</span> <span class="n">name</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;|&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span>
<span class="n">tax2names</span><span class="p">,</span> <span class="n">tax2lineages</span><span class="p">,</span> <span class="n">tax2rank</span> <span class="o">=</span> <span class="n">tree</span><span class="o">.</span><span class="n">annotate_ncbi_taxa</span><span class="p">()</span>

<span class="nb">print</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">to_str</span><span class="p">(</span><span class="n">props</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">,</span> <span class="s2">&quot;sci_name&quot;</span><span class="p">,</span> <span class="s2">&quot;taxid&quot;</span><span class="p">]))</span>
<span class="c1">#                                                             ╭╴9606|protA,Homo sapiens,9606</span>
<span class="c1">#                                  ╭╴(empty),Homininae,207598╶┤</span>
<span class="c1"># ╴(empty),Euarchontoglires,314146╶┤                          ╰╴9598|protA,Pan troglodytes,9598</span>
<span class="c1">#                                  │</span>
<span class="c1">#                                  ╰╴10090|protB,Mus musculus,10090</span>
</pre></div>
</div>
<p>GTDB taxonomy example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">ete4</span> <span class="kn">import</span> <span class="n">PhyloTree</span>

<span class="c1"># Load the whole leaf name as species taxid.</span>
<span class="n">newick</span> <span class="o">=</span> <span class="s1">&#39;((p__Huberarchaeota,f__Korarchaeaceae)d__Archaea,o__Peptococcales);&#39;</span>

<span class="n">tree</span> <span class="o">=</span> <span class="n">PhyloTree</span><span class="p">(</span><span class="n">newick</span><span class="p">)</span>
<span class="n">tax2name</span><span class="p">,</span> <span class="n">tax2track</span><span class="p">,</span> <span class="n">tax2rank</span> <span class="o">=</span> <span class="n">gtdb</span><span class="o">.</span><span class="n">annotate_tree</span><span class="p">(</span><span class="n">tree</span><span class="p">,</span> <span class="n">taxid_attr</span><span class="o">=</span><span class="s2">&quot;name&quot;</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="n">tree</span><span class="o">.</span><span class="n">to_str</span><span class="p">(</span><span class="n">props</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;sci_name&#39;</span><span class="p">,</span> <span class="s1">&#39;rank&#39;</span><span class="p">]))</span>
<span class="c1">#                                         ╭╴p__Huberarchaeota,phylum</span>
<span class="c1">#               ╭╴d__Archaea,superkingdom╶┤</span>
<span class="c1"># ╴root,no rank╶┤                         ╰╴f__Korarchaeaceae,family</span>
<span class="c1">#               │</span>
<span class="c1">#               ╰╴o__Peptococcales,order</span>
</pre></div>
</div>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">ETE Toolkit</a></h1>








<h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../about.html">About</a></li>
<li class="toctree-l1"><a class="reference internal" href="../faqs.html">Frequently Asked Questions (FAQs)</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Tutorial</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="tutorial_trees.html">Working with the Tree structure</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial_phylogeny.html">Phylogenetic trees</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial_smartview.html">Programmable tree drawing</a></li>
<li class="toctree-l2"><a class="reference internal" href="tutorial_drawing.html">Programmable tree drawing</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Taxonomy databases</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../reference/index.html">Reference Guide</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
  <li><a href="index.html">Tutorial</a><ul>
      <li>Previous: <a href="tutorial_drawing.html" title="previous chapter">Programmable tree drawing</a></li>
      <li>Next: <a href="../reference/index.html" title="next chapter">Reference Guide</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>document.getElementById('searchbox').style.display = "block"</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2008 - 2023, The ETE Toolkit Team.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 7.2.6</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.13</a>
      
      |
      <a href="../_sources/tutorial/tutorial_taxonomy.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>