README
=======

.. image:: https://travis-ci.org/jhcepas/ete.svg?branch=master
   :target: https://travis-ci.org/jhcepas/ete

.. image:: https://badges.gitter.im/Join%20Chat.svg
   :alt: Join the chat at https://gitter.im/jhcepas/ete
   :target: https://gitter.im/jhcepas/ete?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge 

..
   .. image:: https://coveralls.io/repos/jhcepas/ete/badge.png


ETE (Environment for Tree Exploration is a Python programming
toolkit that assists in the automated manipulation, analysis and
visualization of phylogenetic trees. Clustering trees or any
other tree-like data structure are also supported. 

ETE is currently developed as a tool for researchers working in phylogenetics
and genomics. ETE offers specialized tools to reconstruct, compare and visualize
phylogenetic trees. If you use ETE for a published work, please cite:

::

  Jaime Huerta-Cepas, Joaquín Dopazo and Toni Gabaldón. ETE: a python
  Environment for Tree Exploration. BMC Bioinformatics 2010, 11:24.


- The official web site of ETE is http://etetoolkit.org
- News and announcements are usually posted on twitter: http://twitter.com/etetoolkit

Gallery
-----------
  
.. image:: https://raw.githubusercontent.com/jhcepas/ete/master/sdoc/gallery.png
   :width: 600
  

Contributing and BUG report
------------------------------
https://github.com/jhcepas/ete/blob/master/CONTRIBUTING.rst


Getting support
------------------

- There is a mailing list providing user support at
  https://groups.google.com/d/forum/etetoolkit. In order to avoid spam,
  messages from new users are moderated. Expect some delay until your first
  message and account is validated.

- For general questions on how to use ETE in bioinformatic projects, the
  BioStars community (http://biostars.org) provides an excellent and broader
  help desk. Please feel free to raise any question there and tag it with the
  "etetoolkit" label.


