.. image:: https://travis-ci.org/etetoolkit/ete.svg?branch=master
   :target: https://travis-ci.org/etetoolkit/ete

.. image:: https://badges.gitter.im/Join%20Chat.svg
   :alt: Join the chat at https://gitter.im/jhcepas/ete
   :target: https://gitter.im/jhcepas/ete?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge 
..
   .. image:: https://coveralls.io/repos/jhcepas/ete/badge.png

.. image:: http://img.shields.io/badge/stackoverflow-etetoolkit-blue.svg
   :target: https://stackoverflow.com/questions/tagged/etetoolkit+or+ete3

.. image:: http://img.shields.io/badge/biostars-etetoolkit-purple.svg
   :target: https://www.biostars.org/t/etetoolkit,ete3,ete,ete2/


Overview
-----------

ETE (Environment for Tree Exploration) is a Python programming toolkit that
assists in the automated manipulation, analysis and visualization of
phylogenetic trees. Clustering trees or any other tree-like data structure are
also supported.

Given that ETE is mainly developed as a tool for researchers working in phylogenetics
and genomics, it also provides specialized tools in that context (e.g. reconstructing, comparing and visualizing
phylogenetic trees). If you use ETE in a published work, please cite:

::

   Jaime Huerta-Cepas, Fran√ßois Serra and Peer Bork. "ETE 3: Reconstruction,
   analysis and visualization of phylogenomic data."  Mol Biol Evol (2016) doi:
   10.1093/molbev/msw046

Install and Documentation
-----------------------------

- The official web site of ETE is http://etetoolkit.org. Downloading
  instructions and further documentation can be found there.

- News and announcements are usually posted on twitter:
  http://twitter.com/etetoolkit

Gallery of examples
--------------------
  
.. image:: https://raw.githubusercontent.com/jhcepas/ete/master/sdoc/gallery.png
   :width: 600
  
Getting Support
------------------
**Please, whenerver possible, avoid sending direct support-related emails to
the developers. Keep communication public:**

- For any type of question on how to use ETE in the bioinformatics context, use BioStars (http://biostars.org) or even StackOverflow forums. 

  Please use the **"etetoolkit"** tag for your questions: 

   .. image:: http://img.shields.io/badge/stackoverflow-etetoolkit-blue.svg
      :target: https://stackoverflow.com/questions/tagged/etetoolkit+or+ete3

   .. image:: http://img.shields.io/badge/biostars-etetoolkit-purple.svg
      :target: https://www.biostars.org/t/etetoolkit,ete3,ete,ete2/

- Bug reports, feature requests and general discussion should be posted into github:
  https://github.com/etetoolkit/ete/issues

- For more technical problems, you can also use the
  official ETE mailing list at https://groups.google.com/d/forum/etetoolkit. To
  avoid spam, messages from new users are moderated. Expect some delay until
  your first message and account is validated.

- For any other inquire (collaborations, sponsoring, etc), please contact *jhcepas /at/ gmail.com*
   

Contributing and BUG reporting
---------------------------------
https://github.com/etetoolkit/ete/wiki/Contributing
  
ROADMAP
--------
https://github.com/etetoolkit/ete/wiki/ROADMAP


