<!DOCTYPE html>

<html lang="en" data-content_root="../">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Taxonomy databases &#8212; ETE Toolkit 4.0.0-beta2 documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=d1102ebc" />
    <link rel="stylesheet" type="text/css" href="../_static/alabaster.css?v=12dfc556" />
    <script src="../_static/documentation_options.js?v=382e6e0c"></script>
    <script src="../_static/doctools.js?v=9a2dae69"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <link rel="icon" href="../_static/favicon.ico"/>
    <link rel="author" title="About these documents" href="../about.html" />
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Smartview (web graphics)" href="reference_smartview.html" />
    <link rel="prev" title="Multiple Sequence Alignments (SeqGroup)" href="reference_seqgroup.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="taxonomy-databases">
<h1>Taxonomy databases<a class="headerlink" href="#taxonomy-databases" title="Link to this heading">¶</a></h1>
<section id="ncbitaxa">
<h2>NCBITaxa<a class="headerlink" href="#ncbitaxa" title="Link to this heading">¶</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="ete4.NCBITaxa">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">NCBITaxa</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dbfile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">taxdump_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">memory</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">update</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/ete4/ncbi_taxonomy/ncbiquery.html#NCBITaxa"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ete4.NCBITaxa" title="Link to this definition">¶</a></dt>
<dd><p>A local transparent connector to the NCBI taxonomy database.</p>
<dl class="py method">
<dt class="sig sig-object py" id="ete4.NCBITaxa.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dbfile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">taxdump_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">memory</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">update</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/ete4/ncbi_taxonomy/ncbiquery.html#NCBITaxa.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ete4.NCBITaxa.__init__" title="Link to this definition">¶</a></dt>
<dd><p>Open and keep a connection to the NCBI taxonomy database.</p>
<p>If it is not present in the system, it will download the
database from the NCBI site first, and convert it to ete’s
format.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ete4.NCBITaxa.annotate_tree">
<span class="sig-name descname"><span class="pre">annotate_tree</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">t</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">taxid_attr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'name'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tax2name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tax2track</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tax2rank</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_unclassified</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/ete4/ncbi_taxonomy/ncbiquery.html#NCBITaxa.annotate_tree"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ete4.NCBITaxa.annotate_tree" title="Link to this definition">¶</a></dt>
<dd><p>Annotate a tree containing taxids as leaf names.</p>
<p>The annotation adds the properties: ‘taxid’, ‘sci_name’,
‘lineage’, ‘named_lineage’ and ‘rank’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>t</strong> – a Tree (or Tree derived) instance.</p></li>
<li><p><strong>taxid_attr</strong> (<em>name</em>) – Allows to set a custom node attribute
containing the taxid number associated to each node (i.e.
species in PhyloTree instances).</p></li>
<li><p><strong>tax2name</strong><strong>,</strong><strong>tax2track</strong><strong>,</strong><strong>tax2rank</strong> – Use these arguments to
provide pre-calculated dictionaries providing translation
from taxid number and names,track lineages and ranks.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ete4.NCBITaxa.get_broken_branches">
<span class="sig-name descname"><span class="pre">get_broken_branches</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">t</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">taxa_lineages</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n2content</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/ete4/ncbi_taxonomy/ncbiquery.html#NCBITaxa.get_broken_branches"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ete4.NCBITaxa.get_broken_branches" title="Link to this definition">¶</a></dt>
<dd><p>Returns a list of NCBI lineage names that are not monophyletic in the
provided tree, as well as the list of affected branches and their size.</p>
<p>CURRENTLY EXPERIMENTAL</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ete4.NCBITaxa.get_common_names">
<span class="sig-name descname"><span class="pre">get_common_names</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">taxids</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/ete4/ncbi_taxonomy/ncbiquery.html#NCBITaxa.get_common_names"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ete4.NCBITaxa.get_common_names" title="Link to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ete4.NCBITaxa.get_descendant_taxa">
<span class="sig-name descname"><span class="pre">get_descendant_taxa</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">parent</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">intermediate_nodes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rank_limit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">collapse_subspecies</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_tree</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/ete4/ncbi_taxonomy/ncbiquery.html#NCBITaxa.get_descendant_taxa"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ete4.NCBITaxa.get_descendant_taxa" title="Link to this definition">¶</a></dt>
<dd><p>Return list of descendant taxids of the given parent.</p>
<p>Parent can be given as taxid or scientific species name.</p>
<p>If intermediate_nodes=True, the list will also have the internal nodes.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ete4.NCBITaxa.get_fuzzy_name_translation">
<span class="sig-name descname"><span class="pre">get_fuzzy_name_translation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sim</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.9</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/ete4/ncbi_taxonomy/ncbiquery.html#NCBITaxa.get_fuzzy_name_translation"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ete4.NCBITaxa.get_fuzzy_name_translation" title="Link to this definition">¶</a></dt>
<dd><p>Return taxid, species name and match score from the NCBI database.</p>
<p>The results are for the best match for name in the NCBI
database of taxa names, with a word similarity &gt;= <cite>sim</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>name</strong> – Species name (does not need to be exact).</p></li>
<li><p><strong>sim</strong> (<em>0.9</em>) – Min word similarity to report a match (from 0 to 1).</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ete4.NCBITaxa.get_lineage">
<span class="sig-name descname"><span class="pre">get_lineage</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">taxid</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/ete4/ncbi_taxonomy/ncbiquery.html#NCBITaxa.get_lineage"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ete4.NCBITaxa.get_lineage" title="Link to this definition">¶</a></dt>
<dd><p>Return lineage track corresponding to the given taxid.</p>
<p>The lineage track is a hierarchically sorted list of parent taxids.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ete4.NCBITaxa.get_lineage_translator">
<span class="sig-name descname"><span class="pre">get_lineage_translator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">taxids</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/ete4/ncbi_taxonomy/ncbiquery.html#NCBITaxa.get_lineage_translator"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ete4.NCBITaxa.get_lineage_translator" title="Link to this definition">¶</a></dt>
<dd><p>Return dict with lineage tracks corresponding to the given taxids.</p>
<p>The lineage tracks are a hierarchically sorted list of parent taxids.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ete4.NCBITaxa.get_name_translator">
<span class="sig-name descname"><span class="pre">get_name_translator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">names</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/ete4/ncbi_taxonomy/ncbiquery.html#NCBITaxa.get_name_translator"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ete4.NCBITaxa.get_name_translator" title="Link to this definition">¶</a></dt>
<dd><p>Return dict with taxids corresponding to the given scientific names.</p>
<p>Exact name match is required for translation.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ete4.NCBITaxa.get_rank">
<span class="sig-name descname"><span class="pre">get_rank</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">taxids</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/ete4/ncbi_taxonomy/ncbiquery.html#NCBITaxa.get_rank"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ete4.NCBITaxa.get_rank" title="Link to this definition">¶</a></dt>
<dd><p>Return dict with NCBI taxonomy ranks for each list of taxids.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ete4.NCBITaxa.get_taxid_translator">
<span class="sig-name descname"><span class="pre">get_taxid_translator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">taxids</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">try_synonyms</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/ete4/ncbi_taxonomy/ncbiquery.html#NCBITaxa.get_taxid_translator"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ete4.NCBITaxa.get_taxid_translator" title="Link to this definition">¶</a></dt>
<dd><p>Return dict with the scientific names corresponding to the taxids.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ete4.NCBITaxa.get_topology">
<span class="sig-name descname"><span class="pre">get_topology</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">taxids</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">intermediate_nodes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rank_limit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">collapse_subspecies</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/ete4/ncbi_taxonomy/ncbiquery.html#NCBITaxa.get_topology"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ete4.NCBITaxa.get_topology" title="Link to this definition">¶</a></dt>
<dd><p>Return the minimal pruned NCBI taxonomy tree containing taxids.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>intermediate_nodes</strong> (<em>False</em>) – If True, single child nodes
representing the complete lineage of leaf nodes are kept.
Otherwise, the tree is pruned to contain the first common
ancestor of each group.</p></li>
<li><p><strong>rank_limit</strong> (<em>None</em>) – If valid NCBI rank name is provided,
the tree is pruned at that given level. For instance, use
rank=”species” to get rid of sub-species or strain leaf
nodes.</p></li>
<li><p><strong>collapse_subspecies</strong> (<em>False</em>) – If True, any item under the
species rank will be collapsed into the species upper
node.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ete4.NCBITaxa.translate_to_names">
<span class="sig-name descname"><span class="pre">translate_to_names</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">taxids</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/ete4/ncbi_taxonomy/ncbiquery.html#NCBITaxa.translate_to_names"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ete4.NCBITaxa.translate_to_names" title="Link to this definition">¶</a></dt>
<dd><p>Return list of scientific names corresponding to taxids.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ete4.NCBITaxa.update_taxonomy_database">
<span class="sig-name descname"><span class="pre">update_taxonomy_database</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">taxdump_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/ete4/ncbi_taxonomy/ncbiquery.html#NCBITaxa.update_taxonomy_database"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ete4.NCBITaxa.update_taxonomy_database" title="Link to this definition">¶</a></dt>
<dd><p>Update the ncbi taxonomy database.</p>
<p>It does it by downloading and parsing the latest
taxdump.tar.gz file from the NCBI site.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>taxdump_file</strong> – Alternative location of the
taxdump.tax.gz file.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="gtdbtaxa">
<h2>GTDBTaxa<a class="headerlink" href="#gtdbtaxa" title="Link to this heading">¶</a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="ete4.GTDBTaxa">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">GTDBTaxa</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dbfile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">taxdump_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">memory</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/ete4/gtdb_taxonomy/gtdbquery.html#GTDBTaxa"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ete4.GTDBTaxa" title="Link to this definition">¶</a></dt>
<dd><p>Local transparent connector to the GTDB taxonomy database.</p>
<dl class="py method">
<dt class="sig sig-object py" id="ete4.GTDBTaxa.__init__">
<span class="sig-name descname"><span class="pre">__init__</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dbfile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">taxdump_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">memory</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/ete4/gtdb_taxonomy/gtdbquery.html#GTDBTaxa.__init__"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ete4.GTDBTaxa.__init__" title="Link to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ete4.GTDBTaxa.annotate_tree">
<span class="sig-name descname"><span class="pre">annotate_tree</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">t</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">taxid_attr</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'name'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tax2name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tax2track</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tax2rank</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_unclassified</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/ete4/gtdb_taxonomy/gtdbquery.html#GTDBTaxa.annotate_tree"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ete4.GTDBTaxa.annotate_tree" title="Link to this definition">¶</a></dt>
<dd><p>Annotate a tree containing taxids as leaf names.</p>
<p>It annotates by adding the properties ‘taxid’, ‘sci_name’,
‘lineage’, ‘named_lineage’ and ‘rank’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>t</strong> – Tree to annotate.</p></li>
<li><p><strong>taxid_attr</strong> – Node attribute (property) containing the
taxid number associated to each node (i.e. species in
PhyloTree instances).</p></li>
<li><p><strong>tax2rank</strong> (<em>tax2name</em><em>, </em><em>tax2track</em><em>,</em>) – Pre-calculated
dictionaries with translations from taxid number to names,
track lineages and ranks.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ete4.GTDBTaxa.get_broken_branches">
<span class="sig-name descname"><span class="pre">get_broken_branches</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">t</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">taxa_lineages</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n2content</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/ete4/gtdb_taxonomy/gtdbquery.html#GTDBTaxa.get_broken_branches"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ete4.GTDBTaxa.get_broken_branches" title="Link to this definition">¶</a></dt>
<dd><p>Returns a list of GTDB lineage names that are not monophyletic in the
provided tree, as well as the list of affected branches and their size.
CURRENTLY EXPERIMENTAL</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ete4.GTDBTaxa.get_common_names">
<span class="sig-name descname"><span class="pre">get_common_names</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">taxids</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/ete4/gtdb_taxonomy/gtdbquery.html#GTDBTaxa.get_common_names"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ete4.GTDBTaxa.get_common_names" title="Link to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ete4.GTDBTaxa.get_descendant_taxa">
<span class="sig-name descname"><span class="pre">get_descendant_taxa</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">parent</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">intermediate_nodes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rank_limit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">collapse_subspecies</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_tree</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/ete4/gtdb_taxonomy/gtdbquery.html#GTDBTaxa.get_descendant_taxa"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ete4.GTDBTaxa.get_descendant_taxa" title="Link to this definition">¶</a></dt>
<dd><p>given a parent taxid or scientific species name, returns a list of all its descendants taxids.
If intermediate_nodes is set to True, internal nodes will also be dumped.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ete4.GTDBTaxa.get_name_lineage">
<span class="sig-name descname"><span class="pre">get_name_lineage</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">taxnames</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/ete4/gtdb_taxonomy/gtdbquery.html#GTDBTaxa.get_name_lineage"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ete4.GTDBTaxa.get_name_lineage" title="Link to this definition">¶</a></dt>
<dd><p>Given a valid taxname, return its corresponding lineage track as a
hierarchically sorted list of parent taxnames.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ete4.GTDBTaxa.get_rank">
<span class="sig-name descname"><span class="pre">get_rank</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">taxids</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/ete4/gtdb_taxonomy/gtdbquery.html#GTDBTaxa.get_rank"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ete4.GTDBTaxa.get_rank" title="Link to this definition">¶</a></dt>
<dd><p>Give a list of GTDB string taxids, return a dictionary with their corresponding ranks.
Examples:</p>
<p>&gt; gtdb.get_rank([‘c__Thorarchaeia’, ‘RS_GCF_001477695.1’])
{‘c__Thorarchaeia’: ‘class’, ‘RS_GCF_001477695.1’: ‘subspecies’}</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ete4.GTDBTaxa.get_topology">
<span class="sig-name descname"><span class="pre">get_topology</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">taxnames</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">intermediate_nodes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rank_limit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">collapse_subspecies</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/ete4/gtdb_taxonomy/gtdbquery.html#GTDBTaxa.get_topology"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ete4.GTDBTaxa.get_topology" title="Link to this definition">¶</a></dt>
<dd><p>Return minimal pruned GTDB taxonomy tree containing all given taxids.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>intermediate_nodes</strong> – If True, single child nodes
representing the complete lineage of leaf nodes are kept.
Otherwise, the tree is pruned to contain the first common
ancestor of each group.</p></li>
<li><p><strong>rank_limit</strong> – If valid NCBI rank name is provided, the
tree is pruned at that given level. For instance, use
rank=”species” to get rid of sub-species or strain leaf
nodes.</p></li>
<li><p><strong>collapse_subspecies</strong> – If True, any item under the
species rank will be collapsed into the species upper
node.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="ete4.GTDBTaxa.update_taxonomy_database">
<span class="sig-name descname"><span class="pre">update_taxonomy_database</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">taxdump_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/ete4/gtdb_taxonomy/gtdbquery.html#GTDBTaxa.update_taxonomy_database"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#ete4.GTDBTaxa.update_taxonomy_database" title="Link to this definition">¶</a></dt>
<dd><p>Update the GTDB taxonomy database.</p>
<p>It updates it by downloading and parsing the latest
gtdbtaxdump.tar.gz file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>taxdump_file</strong> – Alternative location of gtdbtaxdump.tar.gz.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">ETE Toolkit</a></h1>








<h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../about.html">About</a></li>
<li class="toctree-l1"><a class="reference internal" href="../tutorial/index.html">Tutorial</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Reference Guide</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="reference_tree.html">Tree (main class)</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_operations.html">Tree operations</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_parsers.html">Parsers</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_phylo.html">Phylogenetic trees</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_clustering.html">Clustering</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_seqgroup.html">Multiple Sequence Alignments (SeqGroup)</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Taxonomy databases</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_smartview.html">Smartview (web graphics)</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_treeview.html">Treeview (qt graphics)</a></li>
<li class="toctree-l2"><a class="reference internal" href="reference_treematcher.html">Tree matcher</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../faqs.html">Frequently Asked Questions</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
  <li><a href="index.html">Reference Guide</a><ul>
      <li>Previous: <a href="reference_seqgroup.html" title="previous chapter">Multiple Sequence Alignments (SeqGroup)</a></li>
      <li>Next: <a href="reference_smartview.html" title="next chapter">Smartview (web graphics)</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<search id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &#169;2008 - 2023, The ETE Toolkit Team.
      
      |
      Powered by <a href="https://www.sphinx-doc.org/">Sphinx 7.3.7</a>
      &amp; <a href="https://alabaster.readthedocs.io">Alabaster 0.7.16</a>
      
      |
      <a href="../_sources/reference/reference_taxonomy.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>